File: examples/python/basic.py
"""
Basic Zapcode example — execute TypeScript from Python.

Prerequisites: build zapcode-py (see README)
Run with: python examples/python/basic.py
"""

from zapcode import Zapcode, ZapcodeSnapshot

# --- 1. Simple expression ---
b = Zapcode("1 + 2 * 3")
result = b.run()
print("1 + 2 * 3 =", result["output"])  # 7

# --- 2. Using inputs ---
b = Zapcode(
    """
    const greeting = `Hello, ${name}! You are ${age} years old.`;
    greeting
    """,
    inputs=["name", "age"],
)
result = b.run({"name": "Zapcode", "age": 30})
print(result["output"])  # "Hello, Zapcode! You are 30 years old."

# --- 3. Data processing ---
b = Zapcode("""
    const items = [
        { name: "Widget", price: 25.99, qty: 3 },
        { name: "Gadget", price: 49.99, qty: 1 },
        { name: "Doohickey", price: 9.99, qty: 10 },
    ];
    const total = items.reduce((sum, item) => sum + item.price * item.qty, 0);
    const expensive = items.filter(item => item.price > 20).map(i => i.name);
    ({ total, expensive })
""")
result = b.run()
print(result["output"])  # {'total': 227.86, 'expensive': ['Widget', 'Gadget']}

# --- 4. External function (snapshot/resume) ---
b = Zapcode(
    """
    const weather = await getWeather(city);
    const summary = `Weather in ${city}: ${weather.condition}, ${weather.temp}°C`;
    summary
    """,
    inputs=["city"],
    external_functions=["getWeather"],
)

state = b.start({"city": "London"})

if state.get("suspended"):
    print(f"Suspended on: {state['function_name']}({state['args']})")

    # In a real app, call an actual weather API
    mock_weather = {"condition": "Partly cloudy", "temp": 18}

    # Resume with the result
    snapshot = state["snapshot"]
    final = snapshot.resume(mock_weather)
    print(final["output"])  # "Weather in London: Partly cloudy, 18°C"

# --- 5. Resource limits ---
try:
    b = Zapcode("while (true) {}", time_limit_ms=100)
    b.run()
except RuntimeError as e:
    print(f"Caught: {e}")  # allocation limit or time limit

# --- 6. Snapshot serialization ---
b = Zapcode(
    "const data = await fetchData(url); data.length",
    inputs=["url"],
    external_functions=["fetchData"],
)

state = b.start({"url": "https://example.com"})
if state.get("suspended"):
    # Serialize to bytes — store in a database, send over a queue, etc.
    snapshot_bytes = state["snapshot"].dump()
    print(f"Snapshot size: {len(snapshot_bytes)} bytes")

    # Later (possibly in a different process): restore and resume
    restored = ZapcodeSnapshot.load(snapshot_bytes)
    final = restored.resume("hello world")
    print(f"Restored result: {final['output']}")  # 11

File: examples/rust/basic/basic.rs
//! Basic Zapcode example — execute TypeScript from Rust.
//!
//! Run with: cargo run --example basic

use zapcode_core::{ZapcodeRun, ResourceLimits, Value, VmState};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // --- 1. Simple expression ---
    let runner = ZapcodeRun::new(
        "1 + 2 * 3".to_string(),
        vec![],
        vec![],
        ResourceLimits::default(),
    )?;
    let result = runner.run_simple()?;
    println!("1 + 2 * 3 = {:?}", result); // Int(7)

    // --- 2. Using inputs ---
    let runner = ZapcodeRun::new(
        r#"
            const greeting = `Hello, ${name}! You are ${age} years old.`;
            greeting
        "#
        .to_string(),
        vec!["name".to_string(), "age".to_string()],
        vec![],
        ResourceLimits::default(),
    )?;
    let result = runner.run(vec![
        ("name".to_string(), Value::String("Zapcode".into())),
        ("age".to_string(), Value::Int(30)),
    ])?;
    println!("Greeting: {:?}", result.state); // Complete("Hello, Zapcode! You are 30 years old.")

    // --- 3. External function (snapshot/resume) ---
    let runner = ZapcodeRun::new(
        r#"
            const weather = await getWeather(city);
            const summary = `Weather in ${city}: ${weather.condition}, ${weather.temp}°C`;
            summary
        "#
        .to_string(),
        vec!["city".to_string()],
        vec!["getWeather".to_string()],
        ResourceLimits::default(),
    )?;

    // Start execution — suspends at getWeather()
    let state = runner.start(vec![
        ("city".to_string(), Value::String("London".into())),
    ])?;

    match state {
        VmState::Suspended {
            function_name,
            args,
            snapshot,
        } => {
            println!("Suspended on: {}({:?})", function_name, args);

            // In a real app, you'd call an actual weather API here.
            // For this example, we return a mock response.
            let weather_data = Value::Object(indexmap::indexmap! {
                "condition".into() => Value::String("Partly cloudy".into()),
                "temp".into() => Value::Int(18),
            });

            // Resume with the mock result
            let final_state = snapshot.resume(weather_data)?;
            match final_state {
                VmState::Complete(value) => {
                    println!("Result: {:?}", value);
                    // "Weather in London: Partly cloudy, 18°C"
                }
                _ => println!("Unexpected second suspension"),
            }
        }
        VmState::Complete(value) => {
            println!("Completed immediately: {:?}", value);
        }
    }

    // --- 4. Snapshot serialization (store and resume later) ---
    let runner = ZapcodeRun::new(
        r#"
            const data = await fetchData(url);
            data.length
        "#
        .to_string(),
        vec!["url".to_string()],
        vec!["fetchData".to_string()],
        ResourceLimits::default(),
    )?;

    let state = runner.start(vec![
        ("url".to_string(), Value::String("https://example.com".into())),
    ])?;

    if let VmState::Suspended { snapshot, .. } = state {
        // Serialize to bytes — store in a database, send over the network, etc.
        let bytes = snapshot.dump()?;
        println!("Snapshot size: {} bytes", bytes.len());

        // Later (possibly in a different process): restore and resume
        let restored = zapcode_core::ZapcodeSnapshot::load(&bytes)?;
        let final_state = restored.resume(Value::String("hello world".into()))?;
        if let VmState::Complete(value) = final_state {
            println!("Restored result: {:?}", value); // Int(11)
        }
    }

    // --- 5. Async map with multiple external calls ---
    // arr.map(async fn => await external()) now works —
    // each external call suspends/resumes sequentially.
    let runner = ZapcodeRun::new(
        r#"
            const cities = ["London", "Tokyo", "Paris"];
            const results = cities.map(async (city) => {
                const weather = await getWeather(city);
                return weather;
            });
            results
        "#
        .to_string(),
        vec![],
        vec!["getWeather".to_string()],
        ResourceLimits::default(),
    )?;

    let mut state = runner.start(vec![])?;

    // The VM suspends once per city — resolve each one
    let mock_data = vec![
        ("London", "Rainy, 12°C"),
        ("Tokyo", "Clear, 26°C"),
        ("Paris", "Sunny, 22°C"),
    ];

    for (expected_city, weather) in &mock_data {
        match state {
            VmState::Suspended {
                function_name,
                args,
                snapshot,
            } => {
                println!(
                    "  -> {}({}) = {}",
                    function_name,
                    args[0].to_js_string(),
                    weather
                );
                assert_eq!(function_name, "getWeather");
                assert_eq!(args[0].to_js_string(), *expected_city);
                state = snapshot.resume(Value::String((*weather).into()))?;
            }
            VmState::Complete(_) => panic!("expected suspension for {}", expected_city),
        }
    }

    match state {
        VmState::Complete(value) => {
            println!("Async map result: {:?}", value);
            // Array(["Rainy, 12°C", "Clear, 26°C", "Sunny, 22°C"])
        }
        _ => println!("Unexpected suspension after all cities resolved"),
    }

    Ok(())
}
